File: java/src/main/java/com/a65apps/kalimullinilnazrafilovich/interactors/notification/NotificationModel.java
package com.a65apps.kalimullinilnazrafilovich.interactors.notification;

import com.a65apps.kalimullinilnazrafilovich.entities.BirthdayNotification;
import com.a65apps.kalimullinilnazrafilovich.entities.ContactDetailsInfo;
import com.a65apps.kalimullinilnazrafilovich.interactors.calendar.BirthdayCalendar;
import com.a65apps.kalimullinilnazrafilovich.interactors.time.CurrentTime;

import java.util.Calendar;
import java.util.GregorianCalendar;

public class NotificationModel implements NotificationInteractor {
    private static final int FEBRUARY_29 = 29;
    private static final int NEXT_LEAP_YEAR = 4;

    private final transient NotificationRepository notificationRepository;

    private final transient CurrentTime currentTime;
    private final transient GregorianCalendar birthdayGregorianCalendar;

    public NotificationModel(
            NotificationRepository notificationRepository,
            CurrentTime currentTime,
            BirthdayCalendar birthdayCalendar) {
        this.notificationRepository = notificationRepository;
        this.currentTime = currentTime;
        this.birthdayGregorianCalendar = birthdayCalendar.getBirthdayCalendar();
    }

    @Override
    public BirthdayNotification onBirthdayNotification(ContactDetailsInfo contactDetailsInfo) {
        if (birthdayGregorianCalendar.isLeapYear(birthdayGregorianCalendar.get(GregorianCalendar.YEAR))) {
            return setBirthdayReminderForLeapYear(contactDetailsInfo);
        } else {
            return setBirthdayReminder(contactDetailsInfo);
        }
    }

    @Override
    public BirthdayNotification offBirthdayNotification(ContactDetailsInfo contactDetailsInfo) {
        return notificationRepository.removeBirthdayReminder(contactDetailsInfo);
    }

    @Override
    public BirthdayNotification getNotificationWorkStatus(ContactDetailsInfo contactDetailsInfo) {
        return notificationRepository.getBirthdayNotificationEntity(contactDetailsInfo, null);
    }

    private BirthdayNotification setBirthdayReminder(ContactDetailsInfo contactDetailsInfo) {
        GregorianCalendar gregorianCalendar = createGregorianCalendarForContact(contactDetailsInfo);

        if (currentTime.now() > gregorianCalendar.getTime().getTime()
                && !(gregorianCalendar.get(Calendar.MONTH) == Calendar.FEBRUARY
                && gregorianCalendar.get(Calendar.DATE) == FEBRUARY_29)) {
            gregorianCalendar.add(Calendar.YEAR, 1);
        }

        return notificationRepository.setBirthdayReminder(contactDetailsInfo, gregorianCalendar);
    }

    private BirthdayNotification setBirthdayReminderForLeapYear(ContactDetailsInfo contactDetailsInfo) {
        GregorianCalendar gregorianCalendar = createGregorianCalendarForContact(contactDetailsInfo);

        if (currentTime.now() > gregorianCalendar.getTime().getTime()
                && contactDetailsInfo.getDateOfBirth().get(Calendar.MONTH) == Calendar.FEBRUARY
                && contactDetailsInfo.getDateOfBirth().get(Calendar.DAY_OF_MONTH) == FEBRUARY_29) {
            gregorianCalendar.add(Calendar.YEAR, NEXT_LEAP_YEAR);
        } else {
            gregorianCalendar.add(Calendar.YEAR, 1);
        }

        return notificationRepository.setBirthdayReminder(contactDetailsInfo, gregorianCalendar);
    }


    private GregorianCalendar createGregorianCalendarForContact(ContactDetailsInfo contactDetailsInfo) {
        if (contactDetailsInfo.getDateOfBirth().get(Calendar.DAY_OF_MONTH) == FEBRUARY_29) {
            birthdayGregorianCalendar.set(GregorianCalendar.YEAR, birthdayGregorianCalendar.get(Calendar.YEAR));
            while (!birthdayGregorianCalendar.isLeapYear(birthdayGregorianCalendar.get(GregorianCalendar.YEAR))) {
                birthdayGregorianCalendar.add(Calendar.YEAR, 1);
            }
        } else {
            birthdayGregorianCalendar.set(GregorianCalendar.YEAR, birthdayGregorianCalendar.get(Calendar.YEAR));
        }

        birthdayGregorianCalendar.set(GregorianCalendar.DAY_OF_MONTH,
                contactDetailsInfo.getDateOfBirth().get(GregorianCalendar.DAY_OF_MONTH));
        birthdayGregorianCalendar.set(GregorianCalendar.MONTH,
                contactDetailsInfo.getDateOfBirth().get(GregorianCalendar.MONTH));

        return birthdayGregorianCalendar;
    }
}

File: java/src/main/java/com/a65apps/kalimullinilnazrafilovich/interactors/notification/NotificationModel.kt
package com.a65apps.kalimullinilnazrafilovich.interactors.notification

import com.a65apps.kalimullinilnazrafilovich.entities.BirthdayNotification
import com.a65apps.kalimullinilnazrafilovich.entities.ContactDetailsInfo
import com.a65apps.kalimullinilnazrafilovich.interactors.calendar.BirthdayCalendar
import com.a65apps.kalimullinilnazrafilovich.interactors.time.CurrentTime
import java.util.*

class NotificationModel(
        private val notificationRepository: NotificationRepository,
        private val currentTime: CurrentTime,
        birthdayCalendar: BirthdayCalendar
) : NotificationInteractor {
    private val FEBRUARY_29 = 29
    private val NEXT_LEAP_YEAR = 4

    private val birthdayGregorianCalendar = birthdayCalendar.birthdayCalendar

    override fun onBirthdayNotification(contactDetailsInfo: ContactDetailsInfo): BirthdayNotification {
        return if (birthdayGregorianCalendar.isLeapYear(birthdayGregorianCalendar.get(GregorianCalendar.YEAR))) {
            setBirthdayReminderForLeapYear(contactDetailsInfo)
        } else {
            setBirthdayReminder(contactDetailsInfo)
        }
    }

    override fun offBirthdayNotification(contactDetailsInfo: ContactDetailsInfo): BirthdayNotification {
        return notificationRepository.removeBirthdayReminder(contactDetailsInfo)
    }

    override fun getNotificationWorkStatus(contactDetailsInfo: ContactDetailsInfo): BirthdayNotification {
        return notificationRepository.getBirthdayNotificationEntity(contactDetailsInfo, null)
    }

    private fun setBirthdayReminder(contactDetailsInfo: ContactDetailsInfo): BirthdayNotification {
        val gregorianCalendar = createGregorianCalendarForContact(contactDetailsInfo)

        if (currentTime.now() > gregorianCalendar.time.time
                && !(gregorianCalendar.get(Calendar.MONTH) == Calendar.FEBRUARY
                        && gregorianCalendar.get(Calendar.DAY_OF_MONTH) == FEBRUARY_29)) {
            gregorianCalendar.add(Calendar.YEAR, 1)
        }

        return notificationRepository.setBirthdayReminder(contactDetailsInfo, gregorianCalendar)
    }

    private fun setBirthdayReminderForLeapYear(contactDetailsInfo: ContactDetailsInfo): BirthdayNotification {
        val gregorianCalendar = createGregorianCalendarForContact(contactDetailsInfo)

        if (currentTime.now() > gregorianCalendar.time.time
                && contactDetailsInfo.dateOfBirth.get(Calendar.MONTH) == Calendar.FEBRUARY
                && contactDetailsInfo.dateOfBirth.get(Calendar.DAY_OF_MONTH) == FEBRUARY_29) {
            gregorianCalendar.add(Calendar.YEAR, NEXT_LEAP_YEAR)
        } else {
            gregorianCalendar.add(Calendar.YEAR, 1)
        }

        return notificationRepository.setBirthdayReminder(contactDetailsInfo, gregorianCalendar)
    }

    private fun createGregorianCalendarForContact(contactDetailsInfo: ContactDetailsInfo): GregorianCalendar {
        if (contactDetailsInfo.dateOfBirth.get(Calendar.DAY_OF_MONTH) == FEBRUARY_29) {
            birthdayGregorianCalendar.set(GregorianCalendar.YEAR, birthdayGregorianCalendar.get(Calendar.YEAR))
            while (!birthdayGregorianCalendar.isLeapYear(birthdayGregorianCalendar.get(GregorianCalendar.YEAR))) {
                birthdayGregorianCalendar.add(Calendar.YEAR, 1)
            }
        } else {
            birthdayGregorianCalendar.set(GregorianCalendar.YEAR, birthdayGregorianCalendar.get(Calendar.YEAR))
        }

        birthdayGregorianCalendar.set(GregorianCalendar.DAY_OF_MONTH,
                contactDetailsInfo.dateOfBirth.get(GregorianCalendar.DAY_OF_MONTH))
        birthdayGregorianCalendar.set(GregorianCalendar.MONTH,
                contactDetailsInfo.dateOfBirth.get(GregorianCalendar.MONTH))

        return birthdayGregorianCalendar
    }
}